File: stage1/iottymux/iottymux.go
// Copyright 2016 The rkt Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//+build linux

package main

import (
	"bufio"
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"net"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"syscall"
	"time"

	rktlog "github.com/coreos/rkt/pkg/log"
	stage1initcommon "github.com/coreos/rkt/stage1/init/common"

	"github.com/appc/spec/schema/types"
	"github.com/coreos/go-systemd/daemon"
	"github.com/kr/pty"
)

var (
	log     *rktlog.Logger
	diag    *rktlog.Logger
	action  string
	appName string
	debug   bool

	stdinMode  string
	stdoutMode string
	stderrMode string
)

const (
	pathPrefix = "/rkt/iottymux"
)

func init() {
	flag.StringVar(&action, "action", "list", "Sub-action to perform")
	flag.StringVar(&appName, "app", "", "Target application name")
}

type Endpoint struct {
	Name     string `json:"name"`
	Family   string `json:"family"`
	Protocol string `json:"protocol"`
	Address  string `json:"address"`
	Port     string `json:"port"`
	Flavor   string `json:"flavor"`
}

func main() {
	var err error
	// Parse flag and initialize logging
	flag.Parse()
	if os.Getenv("STAGE1_DEBUG") == "true" {
		debug = true
	}
	stage1initcommon.InitDebug(debug)
	log, diag, _ = rktlog.NewLogSet("iottymux", debug)
	if !debug {
		diag.SetOutput(ioutil.Discard)
	}

	// Validate app name
	_, err = types.NewACName(appName)
	if err != nil {
		log.Printf("invalid app name (%s): %v", appName, err)
		os.Exit(254)
	}

	var r error
	switch action {
	case "auto-attach":
		r = actionAttach(true)
	case "custom-attach":
		r = actionAttach(false)
	case "iomux":
		statusFile, e := os.OpenFile(filepath.Join(pathPrefix, appName, "endpoints"), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
		if e != nil {
			err = e
			break
		}
		r = actionIOMux(statusFile)
	case "ttymux":
		statusFile, e := os.OpenFile(filepath.Join(pathPrefix, appName, "endpoints"), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
		if e != nil {
			err = e
			break
		}
		r = actionTTYMux(statusFile)
	case "list":
		fallthrough
	default:
		statusFile, e := os.OpenFile(filepath.Join(pathPrefix, appName, "endpoints"), os.O_RDONLY, os.ModePerm)
		if e != nil {
			err = e
			break
		}
		r = actionPrint(statusFile, os.Stdout)

	}

	if r != nil {
		log.PrintE("runtime failure", r)
		os.Exit(254)
	}
	os.Exit(0)
}

func actionAttach(autoMode bool) error {
	// retrieve available endpoints
	statusFile, err := os.OpenFile(filepath.Join(pathPrefix, appName, "endpoints"), os.O_RDONLY, os.ModePerm)
	if err != nil {
		return err
	}
	eps, err := getEndpoints(statusFile)
	_ = statusFile.Close()
	if err != nil {
		return err
	}

	// retrieve custom attach mode
	var withTTYIn, withTTYOut string
	var withStdin, withStdout, withStderr string
	if !autoMode {
		withTTYIn = os.Getenv("STAGE2_ATTACH_TTYIN")
		withTTYOut = os.Getenv("STAGE2_ATTACH_TTYOUT")
		withStdin = os.Getenv("STAGE2_ATTACH_STDIN")
		withStdout = os.Getenv("STAGE2_ATTACH_STDOUT")
		withStderr = os.Getenv("STAGE2_ATTACH_STDERR")
	}

	ec := make(chan error)
	for s, p := range eps {
		conn, err := net.Dial("tcp", fmt.Sprintf("127.0.0.1:%s", p))
		if err != nil {
			return err
		}
		switch s {
		case "stdin", "tty-in":
			if autoMode || withTTYIn == "true" || withStdin == "true" {
				go io.Copy(conn, os.Stdin)
			}
		case "stdout", "tty-out":
			if autoMode || withTTYOut == "true" || withStdout == "true" {
				go io.Copy(os.Stdout, conn)
			}
		case "stderr":
			if autoMode || withStderr == "true" {
				go io.Copy(os.Stderr, conn)
			}
		case "tty":
			if autoMode || withTTYIn == "true" {
				go io.Copy(conn, os.Stdin)
			}
			if autoMode || withTTYOut == "true" {
				go io.Copy(os.Stdout, conn)
			}
		}
	}
	return <-ec
}

func getEndpoints(in io.Reader) (map[string]string, error) {
	eps := make(map[string]string)
	r := bufio.NewReader(in)
	for {
		line, err := r.ReadString('\n')
		if err == nil {
			fields := strings.Split(line, ",")
			if len(fields) == 4 {
				eps[fields[0]] = strings.Trim(fields[3], "\n")
			}
		}
		if err == io.EOF {
			break
		}
	}
	return eps, nil
}

func actionPrint(in io.Reader, out io.Writer) error {
	eps, err := getEndpoints(in)
	if err != nil {
		return err
	}

	for s, p := range eps {
		out.Write([]byte(fmt.Sprintf("%s available on port %s\n", s, p)))
	}
	return nil
}

func actionTTYMux(statusFile *os.File) error {
	// Create a PTY pair and bind-mount the pts to `/rkt/iottymux/<app>/tty`.
	// Once ready, it will be used by systemd unit.
	pty, tty, err := pty.Open()
	if err != nil {
		return err
	}
	diag.Printf("TTY created at %q", tty.Name())
	ttypath := filepath.Join(pathPrefix, appName, "tty")
	f, err := os.Create(ttypath)
	if err != nil {
		return err
	}
	defer f.Close()
	err = syscall.Mount(tty.Name(), ttypath, "", syscall.MS_BIND, "")
	if err != nil {
		return err
	}

	// signal to systemd that PTY is ready and application can start
	ok, err := daemon.SdNotify("READY=1")
	if !ok {
		return fmt.Errorf("failure during startup notification: %v", err)
	}
	diag.Print("TTY handler ready")

	ttyMode := ""
	if os.Getenv("STAGE2_STDIN") == "true" {
		ttyMode = "tty-in"
	}
	if os.Getenv("STAGE2_STDOUT") == "true" || os.Getenv("STAGE2_STDERR") == "true" {
		if ttyMode == "tty-in" {
			ttyMode = "tty"
		} else {
			ttyMode = "tty-out"
		}

	}
	// Open FIFOs
	var endpoints = make([]*net.Listener, 3)
	var fifos = make([]*os.File, 3)
	if strings.Contains(ttyMode, "tty") {
		fifos[0] = pty
		l, err := net.Listen("tcp4", "127.0.0.1:0")
		if err != nil {
			log.PrintE("unable to open tty listener", err)
			os.Exit(254)
		}
		endpoints[0] = &l
		port := strconv.Itoa((*endpoints[0]).Addr().(*net.TCPAddr).Port)
		diag.Printf("Listening for TTY on %s", port)

		// TODO(lucab): switch to JSON
		ep := strings.Join([]string{ttyMode, "AF_INET4", "127.0.0.1", port}, ",") + "\n"
		_, _ = statusFile.Write([]byte(ep))
		statusFile.Sync()
		statusFile.Close()

	}
	c := make(chan error)
	if fifos[0] != nil && endpoints[0] != nil {
		clients := make(chan *net.Conn)
		go acceptConn(endpoints[0], clients, "tty")
		go proxyIO(clients, fifos[0])
	}
	return <-c
}

func actionIOMux(statusFile *os.File) error {
	var err error
	var endpoints = make([]*net.Listener, 3)
	var fifos = make([]*os.File, 3)

	logMode := os.Getenv("STAGE1_LOGMODE")
	stdinMode = os.Getenv("STAGE2_STDIN")
	stdoutMode = os.Getenv("STAGE2_STDOUT")
	stderrMode = os.Getenv("STAGE2_STDERR")
	// Open FIFOs
	if stdinMode == "true" {
		fifos[0], err = os.OpenFile(filepath.Join(pathPrefix, appName, "stage2-stdin"), os.O_WRONLY, os.ModeNamedPipe)
		if err != nil {
			log.PrintE("invalid stdin FIFO", err)
			os.Exit(254)
		}
		l, err := net.Listen("tcp4", "127.0.0.1:0")
		if err != nil {
			log.PrintE("unable to open stdin listener", err)
			os.Exit(254)
		}
		endpoints[0] = &l

		port := strconv.Itoa((*endpoints[0]).Addr().(*net.TCPAddr).Port)
		ep := strings.Join([]string{"stdin", "AF_INET4", "127.0.0.1", port}, ",") + "\n"
		_, _ = statusFile.Write([]byte(ep))
		diag.Printf("Listening for stdin on %s", port)

	}
	if stdoutMode == "true" {
		fifos[1], err = os.OpenFile(filepath.Join(pathPrefix, appName, "stage2-stdout"), os.O_RDONLY, os.ModeNamedPipe)
		if err != nil {
			log.PrintE("invalid stdout FIFO", err)
			os.Exit(254)
		}
		l, err := net.Listen("tcp4", "127.0.0.1:0")
		if err != nil {
			log.PrintE("unable to open stdout listener", err)
			os.Exit(254)
		}
		endpoints[1] = &l
		port := strconv.Itoa((*endpoints[1]).Addr().(*net.TCPAddr).Port)
		ep := strings.Join([]string{"stdout", "AF_INET4", "127.0.0.1", port}, ",") + "\n"
		_, _ = statusFile.Write([]byte(ep))
		diag.Printf("Listening for stdout on %s", port)

	}
	if stderrMode == "true" {
		fifos[2], err = os.OpenFile(filepath.Join(pathPrefix, appName, "stage2-stderr"), os.O_RDONLY, os.ModeNamedPipe)
		if err != nil {
			log.PrintE("invalid stderr FIFO", err)
			os.Exit(254)
		}
		l, err := net.Listen("tcp4", "127.0.0.1:0")
		if err != nil {
			log.PrintE("unable to open stdout listener", err)
			os.Exit(254)
		}
		endpoints[2] = &l
		port := strconv.Itoa((*endpoints[2]).Addr().(*net.TCPAddr).Port)
		ep := strings.Join([]string{"stderr", "AF_INET4", "127.0.0.1", port}, ",") + "\n"
		_, _ = statusFile.Write([]byte(ep))
		diag.Printf("Listening for stderr on %s", port)

	}
	statusFile.Sync()
	statusFile.Close()
	c := make(chan error)

	var logFile *os.File
	if logMode == "k8s-plain" {
		var err error
		logFile, err = os.OpenFile(filepath.Join(pathPrefix, appName, "logfile"), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
		if err != nil {
			return err
		}
	}
	if fifos[0] != nil && endpoints[0] != nil {
		clients := make(chan *net.Conn)
		go acceptConn(endpoints[0], clients, "stdin")
		go forwardInput(clients, fifos[0])
	}
	if fifos[1] != nil && endpoints[1] != nil {
		localTargets := make(chan io.WriteCloser)
		clients := make(chan *net.Conn)
		lines := make(chan []byte)
		go drainOutput(fifos[1], lines)
		go acceptConn(endpoints[1], clients, "stdout")
		go muxBytes("stdout", lines, clients, localTargets)
		if logFile != nil {
			localTargets <- logFile
		}
	}
	if fifos[2] != nil && endpoints[2] != nil {
		localTargets := make(chan io.WriteCloser)
		clients := make(chan *net.Conn)
		lines := make(chan []byte)
		go drainOutput(fifos[2], lines)
		go acceptConn(endpoints[2], clients, "stderr")
		go muxBytes("stderr", lines, clients, localTargets)
		if logFile != nil {
			localTargets <- logFile
		}
	}

	return <-c
}

func drainOutput(r io.Reader, c chan []byte) {
	rd := bufio.NewReader(r)
	for {
		lineOut, err := rd.ReadBytes('\n')
		if err == nil {
			c <- lineOut
		}
	}
}

func acceptConn(socket *net.Listener, c chan *net.Conn, stream string) {
	for {
		conn, err := (*socket).Accept()
		if err == nil {
			diag.Printf("Accepted new connection for %s", stream)
			c <- &conn
		}
	}
}

func proxyIO(clients chan *net.Conn, tty *os.File) {
	for {
		select {
		case c := <-clients:
			go io.Copy(*c, tty)
			go io.Copy(tty, *c)

		}
	}
}

func forwardInput(clients chan *net.Conn, stdin *os.File) {
	for {
		select {
		case c := <-clients:
			go muxInput(c, stdin)

		}
	}
}

func muxInput(conn *net.Conn, stdin *os.File) {
	rd := bufio.NewReader(*conn)
	for {
		lineIn, err := rd.ReadBytes('\n')
		if err != nil {
			break
		}
		_, err = stdin.Write(lineIn)
		if err != nil {
			break
		}
	}

}

func muxBytes(streamLabel string, lines chan []byte, clients chan *net.Conn, targets chan io.WriteCloser) {
	var logs []io.WriteCloser
	var alivel []bool
	var conns []io.WriteCloser
	var alivec []bool
	for {
		select {
		case l := <-lines:
			for i, f := range logs {
				if alivel[i] == true {
					out := fmt.Sprintf("%s %s %s", time.Now().Format(time.RFC3339Nano), streamLabel, l)
					_, err := f.Write([]byte(out))
					if err != nil {
						f.Close()
						alivel[i] = false
					}
				}
			}
			for i, s := range conns {
				if alivec[i] == true {
					_, err := s.Write(l)
					if err != nil {
						s.Close()
						alivec[i] = false
					}
				}
			}
		case c := <-clients:
			conns = append(conns, *c)
			alivec = append(alivec, true)
		case t := <-targets:
			logs = append(logs, t)
			alivel = append(alivel, true)

		}

	}
}

File: stage1/iottymux/iottymux.c
// Copyright 2016 The rkt Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <errno.h>
#include <stdbool.h>
#include <stdio.h>
#include <string.h>

// fcntl
#include <unistd.h>
#include <fcntl.h>

// mkdir
#include <sys/stat.h>
#include <sys/types.h>

// socket
#include <sys/socket.h>
#include <arpa/inet.h>

// pty
#include <stdlib.h>
#include <fcntl.h>

// tty
#include <termios.h>

// isprint
#include <ctype.h>

// mount
#include <sys/mount.h>

// sd_notify
#include <systemd/sd-daemon.h>

// sd_journal
#include <systemd/sd-journal.h>

// sd event loop
#include <systemd/sd-event.h>
#include <time.h>

static inline void freep(void *p) {
  free(*(void**) p);
}
#define _cleanup_(x) __attribute__((cleanup(x)))
#define _cleanup_free_ _cleanup_(freep)

static int exit_err;
#define exit_if(_cond, _fmt, _args...)              \
  exit_err++;                                       \
  if(_cond) {                                       \
    fprintf(stderr, "Error: " _fmt "\n", ##_args);  \
    exit(exit_err);                                 \
  }

#define pexit_if(_cond, _fmt, _args...)             \
  exit_if(_cond, _fmt ": %s", ##_args, strerror(errno))

#define sdexit_if(_errorno, _fmt, _args...)         \
  exit_if(_errorno < 0, _fmt ": %s", ##_args, strerror(_errorno))


#define TTYDIR "/rkt/tty/"

//TODO(lucab): stacksize/malloc story
#define PTSNAMELEN 512

// ICANON mode
#define BUFLEN 4096
static char in[BUFLEN] = {};
static sd_event *ev = NULL;

#define MAXCLIENTS 3

static struct clients_t {
  int *fd[MAXCLIENTS];
  int len;
} clients = {};

static int tty_out_handler(sd_event_source *src, int fd, uint32_t revents, void *ud) {
  int pty = *((int *) ud);
  int inlen = read(pty, in, BUFLEN-1);
  if (inlen > 0) {
    int i;
    for (i=0; i < clients.len; i++) {
          write(*(clients.fd[i]), in, inlen);
    }
    in[inlen+1] = '\0';
    puts(in);
    /*
    if (in[inlen] != '\n') {
      inlen += putchar('\n');
      fflush(stdout);
    }
    */
  }
  return abs(inlen);
}

// TODO: error handling
static int tcp_handler(sd_event_source *src, int sock, uint32_t revents, void *ud) {
  int pty = *((int *) ud);
  int outlen = 0;
  printf("reading data from client\n");
  int inlen = read(sock, in, BUFLEN-1);
  if (inlen > 0) {
    outlen = write(pty, in, inlen);
  }
  return abs(outlen);
}

// TODO: error handling
static int sock_handler(sd_event_source *src, int sock, uint32_t revents, void *ud) {
  sd_event_source *es = NULL;
  struct sockaddr_in *ca = malloc(sizeof(struct sockaddr_in));
  socklen_t calen = sizeof(struct sockaddr_in);
  bzero(&ca, sizeof(struct sockaddr_in));
  // TODO: cleaning
  int *conn = malloc(sizeof(int));
  *conn = accept(sock, (struct sockaddr *) &ca, &calen);
  if (*conn != -1){
    if (clients.len < MAXCLIENTS) {
      sd_journal_print(LOG_ERR, "Accepted client");
      fcntl(*conn, fcntl(*conn,F_GETFL)|O_NONBLOCK);
      clients.fd[clients.len++] = conn;
      sdexit_if(sd_event_add_io(ev, &es, *conn, EPOLLIN, tcp_handler, ud), "failed to add tcp handler");
    } else {
      close(*conn);
      free(conn);
      sd_journal_print(LOG_ERR, "Too many connected clients, maximum allowed %d", MAXCLIENTS);
    }
  } else {
    sd_journal_print(LOG_ERR, "Failed to accept client");
  }
  return abs(*conn);
}

int main(int argc, char *argv[])
{
  setvbuf (stdout, NULL, _IONBF, 0);
  _cleanup_free_ char* tty_stage1, *dir_stage2, *tty_stage2;
  _cleanup_free_ char* pts_name = malloc(PTSNAMELEN);
  exit_if(pts_name == NULL, "failed to allocate memory for pts name");
  char* app_name = getenv("RKT_APPNAME");
  exit_if(app_name == NULL, "failed to retrieve app name");

  // Setup pty
  struct termios tio;
  int pty = posix_openpt(O_RDWR|O_NOCTTY|O_CLOEXEC|O_NONBLOCK);
  pexit_if(pty == -1, "failed to create pty");
  pexit_if(tcgetattr(pty, &tio) == -1, "failed to set terminal");
  tio.c_lflag &= ~ECHO;
  pexit_if(tcsetattr(pty, TCSANOW, &tio) == -1, "failed to set terminal mode");
  pexit_if(ptsname_r(pty, pts_name, (size_t) PTSNAMELEN) != 0, "failed to get pty slave name");
  pexit_if(grantpt(pty) != 0, "failed to grant pty slave");
  pexit_if(unlockpt(pty) != 0, "failed to unlock pty slave");

  // TODO try using FD_STORE?

  // Bind-mount pts to stage1
  // TODO: conditional mkdir
  int r = asprintf(&tty_stage1, "%s%s", TTYDIR, app_name);
  exit_if(r == -1, "failed to assemble stage1 pts");
  pexit_if(mkdir(TTYDIR, 0755) != 0, "failed to create stage1 tty directory");
  pexit_if(creat(tty_stage1, 0600) == -1, "failed to create stage1 tty %s", tty_stage1);
  pexit_if(mount(pts_name, tty_stage1, NULL, MS_BIND, NULL) != 0, "failed to bindmount pts in stage1");

  // Bind-mount pts to stage2
  // TODO: move to prepare-app
  // TODO: move to /dev/console?
  r = asprintf(&dir_stage2, "/opt/stage2/%s/rootfs/dev/iottymux", app_name);
  exit_if(r == -1, "failed to assemble tty stage1 directory");
  pexit_if(mkdir(dir_stage2, 0755) != 0, "failed to create tty directory");
  r = asprintf(&tty_stage2, "%s/app", dir_stage2);
  exit_if(r == -1, "failed to assemble stage2 tty");
  pexit_if(creat(tty_stage2, 0600) == -1, "failed to create stage2 tty %s", tty_stage2);
  mount(pts_name, tty_stage2, NULL, MS_BIND, NULL);

  // Create socket for remote tty attaching by stage0
  struct sockaddr_in sa;
  bzero(&sa, sizeof(sa));
  sa.sin_family = AF_INET;
  sa.sin_port = 0;
  sa.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
  // TODO: move to AF_VSOCK
  int sock_stage0_tty = socket(AF_INET, SOCK_STREAM|SOCK_NONBLOCK|SOCK_CLOEXEC, 0);
  pexit_if(sock_stage0_tty == -1, "failed to create socket");
  pexit_if(bind(sock_stage0_tty, (struct sockaddr *) &sa, sizeof(sa)) == -1, "failed to bind socket");
  pexit_if(listen(sock_stage0_tty, SOMAXCONN) == -1, "failed to listen on socket");

  // Setup systemd event loop
  sd_event_source *es_ttyout = NULL, *es_sock = NULL;
  sdexit_if(sd_event_default(&ev), "unable to initialize systemd event loop");
  sdexit_if(sd_event_add_io(ev, &es_ttyout, pty, EPOLLIN, tty_out_handler, &pty), "failed to add tty handler");
  sdexit_if(sd_event_source_set_priority(es_ttyout, SD_EVENT_PRIORITY_IMPORTANT), "failed to set tty handler priority");
  sdexit_if(sd_event_add_io(ev, &es_sock, sock_stage0_tty, EPOLLIN, sock_handler, &pty), "failed to add socket handler");

  // Notify readiness, trigger app start
  sd_notify(false, "READY=1");
  sd_journal_print(LOG_DEBUG, "===== Starting log for %s =====", app_name);
  return sd_event_loop(ev);
}
